### PR TITLE
unit test for Elasticsearchconfig.java 

### PR DESCRIPTION
**Pull request by adityaketkar**
## Description
Includes unit test and config files for the test class concerning[ Elasticsearchconfig.java](https://github.com/prasadtalasila/BITS-Darshini/blob/dev/src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java)

## Motivation
Increases the overall code coverage of the project.

## Types of changes
 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Issues Addressed
issue #153

### CODE REVIEW

**Comment by adityaketkar** [comment]
@prasadtalasila  I think the geClient() method for ElasticsearchOperations interface is deprecated.
How can I get getter methods for testing the settings that we have configured?

**Comment by prasadtalasila** [comment]
@adityaketkar Here are a few suggestions for improvement.

1. The PR contains code climate issues. Please correct for them.
1. The PR needs to have a brief description of the changes. You can check an example [PR](https://github.com/DeveloperCAP/SCAT/pull/36).
1. I do not understand your question. Please elaborate.

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java [hunk 1]
```diff
@@ -49,7 +53,19 @@ public ElasticsearchOperations elasticsearchTemplate() {
...
+            .put("index.number_of_shards", 1)
+            .put("index.number_of_replicas", 0)
+            .put("http.cors.enabled",
+                    Boolean.valueOf(corsEnabled).booleanValue())
+            .put("http.cors.allow-origin", allowOrigin)
+            .put("http.cors.allow-methods", allowMethods)
+            .put("http.cors.allow-headers", allowHeaders);
+	NodeBuilder builder = esFactoryImpl.nodeBuilder().local(true)
+			.settings(settingsBuilder.build());
         return new ElasticsearchTemplate(builder.node().client());
```
Comment (adityaketkar) [inline src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java]: @prasadtalasila We configure ElasticsearchTemplate by passing builder.node().client() into the constructor, but there is no getter defined in ElasticsearchTemplate to get this back and test it for mocked inputs.

**Comment by adityaketkar** [comment]
@prasadtalasila Also, I am not very sure whether my approach is correct for the test, can you please review the code?
Once the review is done with, I will optimize the code for all codeclimate issues.

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java (hunk 1 above)
Comment (prasadtalasila) [reply src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java]: You can push the creation of this object into the abstract factory implementation. That way, the problem gets handled by the mocks.

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java [hunk 2]
```diff
@@ -49,7 +53,19 @@ public ElasticsearchOperations elasticsearchTemplate() {
                 .put("http.cors.allow-methods", allowMethods)
                 .put("http.cors.allow-headers", allowHeaders);
         NodeBuilder builder = NodeBuilder.nodeBuilder().local(true)
-                .settings(settingsBuilder.build());
+                .settings(settingsBuilder.build());*/
+	ImmutableSettings.Builder settingsBuilder = esFactoryImpl.settingsBuilder().put("cluster.name", clusterName)
```
Comment (prasadtalasila) [inline src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java]: All of these settings can be pushed to the ESFactoryImpl.settingsBuilder() method. An invocation of that method exists here to represent the idea of creating the setting builder. This chain invocation disappears into the ESFactoryImpl.settingsBuilder() method.

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java (hunk 1 above)
Comment (prasadtalasila) [reply src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java]: You can create create ESFactory.elasticsearchTemplate() method to handle the object returned in the last statement.

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
 import org.elasticsearch.common.settings.ImmutableSettings;
```
Comment (prasadtalasila) [inline src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java]: This class name should be Elastic**S**earchConfig.java

File: src/test/java/unit/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfigTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+		}
+
+		@Test
+		public void wiringTest() {
+			assertThat(elasticSearchConfig, is(notNullValue()));
+		}
+
+		@Test
+		public void testElasticsearchTemplate() {
+	        when(this.environment.getProperty("elasticsearch.cluster.name")).thenReturn("testClusterName");
```
Comment (prasadtalasila) [inline src/test/java/unit/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfigTest.java]: These will go away once the settings are pushed to settingsBuilder() method.

**Review by prasadtalasila — CHANGES_REQUESTED** [review]

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ESFactoryImpl.java [hunk 5]
```diff
@@ -1,25 +1,61 @@
...
 import org.springframework.stereotype.Component;
 
 @Component
 @Scope("prototype")
+@EnableElasticsearchRepositories(basePackages = "in.ac.bits.protocolanalyzer.persistence.repository")
+@PropertySource(value = "classpath:META-INF/elasticsearch.properties")
 public class ESFactoryImpl implements ESFactory {
 
+	@Resource
+    private Environment environment;
```
Comment (adityaketkar) [inline src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ESFactoryImpl.java]: @prasadtalasila when I try to access Environment using getProperty(), it throws a NullPointerException.
I checked and found that field environment is consistently null in this case, but I could not figure out why.
Please help. I will refactor and optimize code once it is in running condition.

**Comment by prasadtalasila** [comment]
@adityaketkar I am still looking through a possible reason, but the most likely problem is the interaction of the @resource annotation with the class-level annotations.

**Comment by adityaketkar** [comment]
@prasadtalasila should I let the configuration involving Environment be in ElasticSearchConfig.java for now?
I was unable to find any solution for the problem after much search.

**Comment by prasadtalasila** [comment]
@adityaketkar Another option is to keep the environment where it is and pass it as an argument to the function in the factory. That way, at least the coupled method invocations move out of the ElasticSearchConfig class.

**Comment by prasadtalasila** [comment]
@adityaketkar It would still be useful to construct a minimum working example that has the problem you are describing. We can find a solution to it and apply the solution to all the similar scenarios.

**Comment by adityaketkar** [comment]
@prasadtalasila I tried invoking the 

     environment.getProperty("elasticsearch.cluster.name") 

in your [dev](https://github.com/prasadtalasila/BITS-Darshini/blob/fafa9e76049bd5feda1986b56cc194bb7d3a462c/src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticsearchConfig.java#L25) branch. It's surprising that this call returns null there too. We never knew about it maybe because there is no call to this function in the code. Can you please confirm ?

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticSearchConfig.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.data.elasticsearch.core.ElasticsearchOperations;
+import org.springframework.data.elasticsearch.core.ElasticsearchTemplate;
+import org.springframework.data.elasticsearch.repository.config.EnableElasticsearchRepositories;
+
+@Configuration
+@EnableElasticsearchRepositories(basePackages = 
+		"in.ac.bits.protocolanalyzer.persistence.repository")
+public class ElasticSearchConfig {
+    
+	@Autowired
```
Comment (prasadtalasila) [inline src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticSearchConfig.java]: @adityaketkar The line indentation is not uniform across the file. Please adjust.

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticSearchFactoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+		String clusterName = env.getProperty("elasticsearch.cluster.name");
+        String nodeName = env.getProperty("elasticsearch.node.name");
+        String corsEnabled = env.getProperty("elasticsearch.http.cors.enabled");
+        String allowOrigin = env
+                .getProperty("elasticsearch.http.cors.allow-origin");
+        String allowMethods = env
+                .getProperty("elasticsearch.http.cors.allow-methods");
+        String allowHeaders = env
+                .getProperty("elasticsearch.http.cors.allow-headers");
+		String dataPath = env.getProperty("elasticsearch.path.data");
```
Comment (prasadtalasila) [inline src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticSearchFactoryImpl.java]: Again the line indentations are not uniform.

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticSearchFactoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+		String logPath = env
+		.getProperty("elasticsearch.path.logs");
+		return ImmutableSettings
+                .settingsBuilder().put("cluster.name",clusterName)
```
Comment (prasadtalasila) [inline src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticSearchFactoryImpl.java]: The methods in the method chain must align at one indentation.

File: src/main/resources/META-INF/elasticsearch.properties [hunk 9]
```diff
@@ -2,7 +2,7 @@
 
 elasticsearch.host = 127.0.0.1
 elasticsearch.port = 9200
-elasticsearch.cluster.name = protocol-analyzer
+elasticsearch.cluster.name=protocol-analyzer
```
Comment (prasadtalasila) [inline src/main/resources/META-INF/elasticsearch.properties]: please revert the change. All other lines have a whitespace before and after equals symbol.

**Review by prasadtalasila — CHANGES_REQUESTED** [review]

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticSearchConfig.java (hunk 6 above)
Comment (adityaketkar) [reply src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticSearchConfig.java]: @prasadtalasila the newline is to prevent the sentence from crossing the 100 character limit imposed by codeclimate

File: src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticSearchFactoryImpl.java (hunk 8 above)
Comment (adityaketkar) [reply src/main/java/in/ac/bits/protocolanalyzer/persistence/repository/ElasticSearchFactoryImpl.java]: Corrected in the latest commit ed2996c

File: src/main/resources/META-INF/elasticsearch.properties (hunk 9 above)
Comment (adityaketkar) [reply src/main/resources/META-INF/elasticsearch.properties]: Corrected in the latest commit ed2996c
